flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java

```diff
@@ -39,6 +44,10 @@ public class ErrorInfo implements Serializable {
 
     private final long timestamp;
 
+    private final transient CompletableFuture<Map<String, String>> labelsFuture;
+    /** Labels associated with the exception, set as soon as labelsFuture is completed. */
+    private Map<String, String> labels;
```

Comment:
🤔 I'd prefer introducing a SerializableCompletableFuture instead

Reply:
haha, this seems to go against https://github.com/apache/flink/pull/22506#discussion_r1192155460

I guess it's a matter of taste, your call. We have to avoid blocking in both cases (which is what @zhuzhurk's comment seems to be mostly about).

Reply:
we should just make sure labels field is marked as volatile, because it could be set from a different thread

Reply:
Sure, extended `FailureEnricherUtils#labelFailure` method to support that